gms/endpoint_state.hh

```diff
@@ -168,4 +168,15 @@ public:
     friend std::ostream& operator<<(std::ostream& os, const endpoint_state& x);
 };
 
+// The endpoint state is protected with an endpoint lock
+// acquired in the gossiper using gossiper::lock_endpoint.
+//
+// permit_id identifies the held endpoint lock
+// and it is used by gossiper::lock_endpoint to prevent a deadlock
+// when a notification function is called under the endpoint lock
+// and calls back into the gossiper on a path that wants to acquire
+// the endpoint_lock for the same endpoint.
+using permit_id = utils::tagged_uuid<struct permit_id_tag>;
+constexpr permit_id null_permit_id = permit_id::create_null_id();
```

Comment:
It's not bypassing the lock but rather validating (with non-null permit_id) that the lock is held for the given endpoint, on a nested call. Today there's either no lock taken (where it should) or you'd hit a deadlock.
Providing the permit_id just prevents this deadlock.

Reply:
Why do we lock at such fine granularity?

Reply:
commit 8f01e08690926b91e7cadb1d33c09e1b1c3536bf
Author: Tomasz Grabiec <tgrabiec@scylladb.com>
Date:   Tue Oct 10 19:07:36 2017 +0200

    gms/gossiper: Serialize state changes and notifications for given node
    
    It's possible that a change listener for a later state will run before
    change listener for the previous state completes, in which case
    node's state can be corruped. For example, the previous change listener
    may override sysytem.peers with an old value.
    
    This patch fixes the problem by serializing state changes and
    listeners for each node.
    
    The implementation uses loading_shared_values so that the lock remains
    alive as long as there is anyone holding it. Using endpoint_state_map
    for that doesn't seem appropraite, because entries can be removed from
    it while listeners are still running.  There is code in the gossiper
    which anticipates that entry may be gone across deferring points in
    some places.


but is it merited? what are the lock hold times? why not lock everything? or require atomicity? The complexity is high.

Reply:
I'm not sure what will it buy us beyond getting rid of the loading_shared_values endpoint_locks_map.
If we lock everything, and hold that lock across notifications, we'll need a similar mechanism to prevent deadlock and will need to cover any endpoint change, like in `handle_state_normal` that calls `remove_endpoint` for other endpoints (not the one it was called for).

Reply:
@tgrabiec what do you think?

Reply:
After this permit_id thing, the locking complexity is too high! Perhaps nobody is going to understand it except Benny.

Reply:
I no longer remember all the details, and they're probably obsolete now anyway, but it seems to me that locking globally has much more potential for deadlocks, like Benny mentioned.